### ISSUE TITLE
slim.batch_norm used with slim.conv2d problem

### ISSUE THREAD

**Issue by ColaWithIce** [tensorflow/tensorflow#5663]
Dear All,
I have met some problems when using the batch norm layer of slim. I trained the same model structure but with different ways to use batch_norm layer, shown below. It seems the output of the two ways to use batch_norm is different to me. I have look inside the code to see. It seems the two way no different at all. But the result is actually not the same. Anyone met it before?

1. net = slim.conv2d(net, 64, [4, 4], 2, normalizer_fn=None, activation_fn=None, biases_initializer=None, reuse=reuse)
   output1 = slim.batch_norm(net)

2. output2 = slim.conv2d(net, 64 ,[4, 4], 2, normalizer_fn=slim.batch_norm, activation_fn=None, reuse=reuse)

**Comment by girving** [comment]
I believe one of those ways normalizes before the activation.  The other does not.

**Comment by mikowals** [comment]
There is no activation function because it is set to None.

Variables aren't being reused in the batch_norm of example 1.  Otherwise I would expect the results to match.

**Comment by girving** [comment]
Can you give a reproducible example?  Even better, if you have a link to the code you think is wrong, can you provide it?  It sounds like you may already know what the problem is.

**Comment by ColaWithIce** [comment]
@mikowals @girving Actually, I read the [code]( https://github.com/tensorflow/tensorflow/blob/r0.11/tensorflow/contrib/layers/python/layers/layers.py), from line 439 to 455, it describes how the conv2d layer uses the batch_norm and activation. What I understand from this is that the `reuse` parameter only influnence the `conv2d` layer but not the function used inside.
To make the two ways same, I also set `biases_initializer` None.

@girving Sorry, I don't know what is the exact problem. But I guess it ways to use `batch_norm`.  I just try one [tutorial](https://github.com/awjuliani/TF-Tutorials/blob/master/DCGAN.ipynb) written by @awjuliani. Here are the two ways I used in the code which I use Ipython notebook to test. What I just change is the `dis3` of the `discriminator`. The first way ipython notebook is shown [here.](https://github.com/ColaWithIce/TF-test/blob/master/DCGAN-Copy1.ipynb) The second is [here.](https://github.com/ColaWithIce/TF-test/blob/master/DCGAN.ipynb) The output of the training message is very different. The two way's result should be similar but now as you can see the Disc Loss are not in same scale. I just wonder why this will happen. Is there something wrong with the variable updating or the way I use the code is wrong.

**Comment by mikowals** [comment]
I made a simpler reproduction and it shows both examples are equivalent.  There could still be a problem with variable reuse if you are not placing example 1 above into a variable_scope with reuse set to true.  

    import tensorflow as tf
    import tensorflow.contrib.slim as slim

    data = tf.random_normal(shape=(64, 16, 16, 3), seed=10)

    conv_bn = slim.conv2d(data, 16, (3,3), activation_fn=None,
                            normalizer_fn=slim.batch_norm, normalizer_params={'is_training': True},
                            scope='conv')

    conv_no_bn = slim.conv2d(data, 16, (3,3), activation_fn=None, biases_initializer=None,
                            reuse=True, scope='conv')
    alt_conv_bn = slim.batch_norm(conv_no_bn, is_training=True)

    is_equal = tf.reduce_all(tf.equal(conv_bn, alt_conv_bn))
    
    moving_variances = tf.contrib.framework.get_variables_by_suffix('moving_variance')
    variance_is_equal = tf.reduce_all(tf.equal(*moving_variances))

    moving_means = tf.contrib.framework.get_variables_by_suffix('moving_mean')
    mean_is_equal = tf.reduce_all(tf.equal(*moving_means))

    with tf.Session() as sess:
      sess.run(tf.initialize_all_variables())
      print(sess.run(is_equal)) # True
      print(sess.run(is_equal)) # True
      print([v.name for v in moving_variances]) # ['conv/BatchNorm/moving_variance:0', 'BatchNorm/moving_variance:0']
      print([v.name for v in moving_means]) # ['conv/BatchNorm/moving_mean:0', 'BatchNorm/moving_mean:0']
      print(sess.run([variance_is_equal, mean_is_equal])) # [True, True]

**Comment by ColaWithIce** [comment]
@mikowals  Thank you for your comment. I think I find the problem. The trainable variable numbers are different between different ways. That means if I use `reuse` in `conv2d` layer, all the variable created in this layer INCLUDE the variables created in `batch_norm` layer will be reused when `reuse` is `True`
What I test is shown below.
For example 1:
```
for v in tf.trainable_variables():
    print 'name = {}'.format(v.value())
    print len(tf.trainable_variables())
```
OUTPUT:
```
name = Tensor("g_project/weights/read:0", shape=(100, 4096), dtype=float32)
name = Tensor("g_project/BatchNorm/beta/read:0", shape=(4096,), dtype=float32)
name = Tensor("g_conv1/weights/read:0", shape=(5, 5, 64, 256), dtype=float32)
name = Tensor("g_conv1/BatchNorm/beta/read:0", shape=(64,), dtype=float32)
name = Tensor("g_conv2/weights/read:0", shape=(5, 5, 32, 64), dtype=float32)
name = Tensor("g_conv2/BatchNorm/beta/read:0", shape=(32,), dtype=float32)
name = Tensor("g_conv3/weights/read:0", shape=(5, 5, 16, 32), dtype=float32)
name = Tensor("g_conv3/BatchNorm/beta/read:0", shape=(16,), dtype=float32)
name = Tensor("g_out/weights/read:0", shape=(32, 32, 1, 16), dtype=float32)
name = Tensor("d_conv1/weights/read:0", shape=(4, 4, 1, 16), dtype=float32)
name = Tensor("d_conv2/weights/read:0", shape=(4, 4, 16, 32), dtype=float32)
name = Tensor("d_conv2/BatchNorm/beta/read:0", shape=(32,), dtype=float32)
name = Tensor("d_conv3/weights/read:0", shape=(4, 4, 32, 64), dtype=float32)
name = Tensor("BatchNorm/beta/read:0", shape=(64,), dtype=float32)
name = Tensor("d_out/weights/read:0", shape=(1024, 1), dtype=float32)
name = Tensor("d_out/biases/read:0", shape=(1,), dtype=float32)
name = Tensor("BatchNorm_1/beta/read:0", shape=(64,), dtype=float32)
17
```

For example 2:
```
for v in tf.all_variables():
    print 'name = {}'.format(v.value())
    print len(tf.all_variables())

```
OUTPUT:
```
name = Tensor("g_project/weights/read:0", shape=(100, 4096), dtype=float32)
name = Tensor("g_project/BatchNorm/beta/read:0", shape=(4096,), dtype=float32)
name = Tensor("g_conv1/weights/read:0", shape=(5, 5, 64, 256), dtype=float32)
name = Tensor("g_conv1/BatchNorm/beta/read:0", shape=(64,), dtype=float32)
name = Tensor("g_conv2/weights/read:0", shape=(5, 5, 32, 64), dtype=float32)
name = Tensor("g_conv2/BatchNorm/beta/read:0", shape=(32,), dtype=float32)
name = Tensor("g_conv3/weights/read:0", shape=(5, 5, 16, 32), dtype=float32)
name = Tensor("g_conv3/BatchNorm/beta/read:0", shape=(16,), dtype=float32)
name = Tensor("g_out/weights/read:0", shape=(32, 32, 1, 16), dtype=float32)
name = Tensor("d_conv1/weights/read:0", shape=(4, 4, 1, 16), dtype=float32)
name = Tensor("d_conv2/weights/read:0", shape=(4, 4, 16, 32), dtype=float32)
name = Tensor("d_conv2/BatchNorm/beta/read:0", shape=(32,), dtype=float32)
name = Tensor("d_conv3/weights/read:0", shape=(4, 4, 32, 64), dtype=float32)
name = Tensor("d_conv3/BatchNorm/beta/read:0", shape=(64,), dtype=float32)
name = Tensor("d_out/weights/read:0", shape=(1024, 1), dtype=float32)
name = Tensor("d_out/biases/read:0", shape=(1,), dtype=float32)
16
```

@mikowals By the way, when I use `slim.batch_norm` alone. Is it the right way to set parameters of normalization {is_training=True} in Training phase and {is_training=False, reuse=True} in Testing and validation phase? My point is do I need to set `reuse` to True in the Testing phase. Or just set the `reuse` as default `None.` Thank you in advance.

**Comment by mikowals** [comment]
Yes, setting `reuse=True` in `slim.conv2d` will also insure the batch norm variables are reused.   So in general your example 2 is the simple way.  

You are also correct with how `{'is_training': True} ` is used with batch norm.  Setting `reuse=True` for testing is also probably correct though often it is best just to save the model and have the evaluation code create its own model and load the variables from a checkpoint.   See examples in [slim.evaluation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/python/slim/evaluation.py).

If your issue is resolved you should close it.
